File: asherah-java/java/src/main/java/com/godaddy/asherah/appencryption/SessionJsonImpl.java
package com.godaddy.asherah.appencryption;

import com.godaddy.asherah.jni.AsherahSession;
import org.json.JSONObject;

import java.nio.charset.StandardCharsets;

/**
 * Session implementation for JSONObject payloads with byte[] data row records.
 * Wraps the native AsherahSession. Compatible with the canonical godaddy/asherah SDK.
 *
 * @param <D> the data row record type (byte[] or JSONObject)
 */
class SessionJsonImpl<D> implements Session<JSONObject, D> {

    private final AsherahSession inner;
    private final boolean drrIsJson;

    SessionJsonImpl(final AsherahSession inner, final boolean drrIsJson) {
        this.inner = inner;
        this.drrIsJson = drrIsJson;
    }

    @Override
    @SuppressWarnings("unchecked")
    public D encrypt(final JSONObject payload) {
        final byte[] data = payload.toString().getBytes(StandardCharsets.UTF_8);
        final String drrJson = inner.encryptToJson(data);
        if (drrIsJson) {
            return (D) new JSONObject(drrJson);
        }
        return (D) drrJson.getBytes(StandardCharsets.UTF_8);
    }

    @Override
    public JSONObject decrypt(final D dataRowRecord) {
        final String drrJson;
        if (dataRowRecord instanceof JSONObject) {
            drrJson = ((JSONObject) dataRowRecord).toString();
        } else {
            drrJson = new String((byte[]) dataRowRecord, StandardCharsets.UTF_8);
        }
        final byte[] plaintext = inner.decryptFromJson(drrJson);
        return new JSONObject(new String(plaintext, StandardCharsets.UTF_8));
    }

    @Override
    public void close() {
        inner.close();
    }
}

File: asherah-dotnet/GoDaddy.Asherah.AppEncryption.Compat/AppEncryption/SessionJsonImpl.cs
using System.Text;
using Newtonsoft.Json.Linq;

namespace GoDaddy.Asherah.AppEncryption;

internal class SessionJsonImpl<TD> : Session<JObject, TD>
{
    private readonly AsherahSession _inner;
    private readonly bool _drrIsJson;

    internal SessionJsonImpl(AsherahSession inner, bool drrIsJson)
    {
        _inner = inner;
        _drrIsJson = drrIsJson;
    }

    public override TD Encrypt(JObject payload)
    {
        var data = Encoding.UTF8.GetBytes(payload.ToString(Newtonsoft.Json.Formatting.None));
        var drrJson = Encoding.UTF8.GetString(_inner.EncryptBytes(data));
        if (_drrIsJson)
            return (TD)(object)JObject.Parse(drrJson);
        return (TD)(object)Encoding.UTF8.GetBytes(drrJson);
    }

    public override JObject Decrypt(TD dataRowRecord)
    {
        var drrJson = dataRowRecord is JObject jobj
            ? jobj.ToString(Newtonsoft.Json.Formatting.None)
            : Encoding.UTF8.GetString((byte[])(object)dataRowRecord!);
        var plaintext = _inner.DecryptBytes(Encoding.UTF8.GetBytes(drrJson));
        return JObject.Parse(Encoding.UTF8.GetString(plaintext));
    }

    public override void Dispose() => _inner.Dispose();
}
